src/artemis/experiment_plans/oav_grid_detection_plan.py

```diff
@@ -0,0 +1,187 @@
+
+    Args:
+        oav (OAV): The OAV device in use.
+        parameters (OAVParamaters): Object containing values loaded in from various parameter files in use.
+        max_run_num (int): Maximum number of times to run.
+        rotation_points (int): Test to see if the pin is widest `rotation_points` number of times on a full 180 degree rotation.
+    """
+    oav: OAV = i03.oav()
+    smargon: Smargon = i03.smargon()
+    LOGGER.info("OAV Centring: Starting loop centring")
+
+    yield from bps.wait()
+
+    # Set relevant PVs to whatever the config dictates.
+    yield from pre_centring_setup_oav(oav, parameters)
+
+    LOGGER.info("OAV Centring: Camera set up")
+
+    start_positions = []
+    box_numbers = []
+
+    box_size_x_pixels = box_size_um / parameters.micronsPerXPixel
+    box_size_y_pixels = box_size_um / parameters.micronsPerYPixel
+
+    for angle in [0, 90]:
+        yield from bps.mv(smargon.omega, angle)
+        # need to wait for the OAV image to update
+        # TODO improve this from just waiting some random time
+        yield from bps.sleep(0.5)
+
+        top_edge = np.array((yield from bps.rd(oav.mxsc.top)))
+        bottom_edge = np.array((yield from bps.rd(oav.mxsc.bottom)))
+
+        tip_x_px = yield from bps.rd(oav.mxsc.tip_x)
+        tip_y_px = yield from bps.rd(oav.mxsc.tip_y)
+
+        LOGGER.info(f"Tip is at x,y: {tip_x_px},{tip_y_px}")
+
+        full_oav_image_height_px = yield from bps.rd(oav.cam.array_size.array_size_y)
+
+        # only use the area from the start of the pin onwards
+        top_edge = top_edge[tip_x_px : tip_x_px + grid_width_px]
+        bottom_edge = bottom_edge[tip_x_px : tip_x_px + grid_width_px]
+
+        # the edge detection line can jump to the edge of the image sometimes, filter
+        # those points out
+        min_y = np.min(top_edge[top_edge != 0])
+        max_y = np.max(bottom_edge[bottom_edge != full_oav_image_height_px])
+        LOGGER.info(f"Min/max {min_y, max_y}")
+        # if top and bottom empty after filter use the whole image (ask neil)
+        grid_height_px = max_y - min_y
+
+        LOGGER.info(f"Drawing snapshot {grid_width_px} by {grid_height_px}")
+
+        boxes = (
+            math.ceil(grid_width_px / box_size_x_pixels),
+            math.ceil(grid_height_px / box_size_y_pixels),
+        )
+        box_numbers.append(boxes)
+
+        upper_left = (tip_x_px, min_y)
+
+        yield from bps.abs_set(oav.snapshot.top_left_x, upper_left[0])
+        yield from bps.abs_set(oav.snapshot.top_left_y, upper_left[1])
+        yield from bps.abs_set(oav.snapshot.box_width, box_size_x_pixels)
+        yield from bps.abs_set(oav.snapshot.num_boxes_x, boxes[0])
+        yield from bps.abs_set(oav.snapshot.num_boxes_y, boxes[1])
+
+        LOGGER.info("Triggering snapshot")
+
+        snapshot_filename = (
+            filenames["snap_1_filename"] if angle == 0 else filenames["snap_2_filename"]
```

Comment:
Will be addressed in https://github.com/DiamondLightSource/python-artemis/issues/651